sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,1632 @@
+        its initial value and not optimized.
+    """
+    def __init__(self, l=1.0, l_bounds=(1e-5, 1e5), nu=1.5):
+        super(Matern, self).__init__(l, l_bounds)
+        self.nu = nu
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """Return the kernel k(X, Y) and optionally its gradient.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples_X, n_features)
+            Left argument of the returned kernel k(X, Y)
+
+        Y : array, shape (n_samples_Y, n_features), (optional, default=None)
+            Right argument of the returned kernel k(X, Y). If None, k(X, X)
+            if evaluated instead.
+
+        eval_gradient : bool (optional, default=False)
+            Determines whether the gradient with respect to the kernel
+            hyperparameter is determined. Only supported when Y is None.
+
+        Returns
+        -------
+        K : array, shape (n_samples_X, n_samples_Y)
+            Kernel k(X, Y)
+
+        K_gradient : array (opt.), shape (n_samples_X, n_samples_X, n_dims)
+            The gradient of the kernel k(X, X) with respect to the
+            hyperparameter of the kernel. Only returned when eval_gradient
+            is True.
+        """
+        X = np.atleast_2d(X)
+        if np.iterable(self.l) and X.shape[1] != self.l.shape[0]:
+            raise Exception("Anisotropic kernel must have the same number of "
+                            "dimensions as data (%d!=%d)"
+                            % (self.l.shape[0], X.shape[1]))
+
+        if Y is None:
+            dists = pdist(X / self.l, metric='euclidean')
+        else:
+            if eval_gradient:
+                raise ValueError(
+                    "Gradient can only be evaluated when Y is None.")
+            dists = cdist(X / self.l, Y / self.l, metric='euclidean')
+
+        if self.nu == 0.5:
+            K = np.exp(-dists)
+        elif self.nu == 1.5:
+            K = dists * math.sqrt(3)
+            K = (1. + K) * np.exp(-K)
+        elif self.nu == 2.5:
+            K = dists * math.sqrt(5)
+            K = (1. + K + K ** 2 / 3.0) * np.exp(-K)
+        else:  # general case; expensive to evaluate
+            K = dists
+            K[K == 0.0] += np.finfo(float).eps  # strict zeros result in nan
+            tmp = (math.sqrt(2 * self.nu) * K)
+            K.fill((2 ** (1. - self.nu)) / gamma(self.nu))
+            K *= tmp ** self.nu
+            K *= kv(self.nu, tmp)
+
+        if Y is None:
+            # convert from upper-triangular matrix to square matrix
+            K = squareform(K)
+            np.fill_diagonal(K, 1)
+
+        if eval_gradient:
+            if self.l_bounds is "fixed":  # Hyperparameter l kept fixed
+                K_gradient = np.empty((X.shape[0], X.shape[0], 0))
+                return K, K_gradient
+
+            # We need to recompute the pairwise dimension-wise distances
+            D = (X[:, np.newaxis, :] - X[np.newaxis, :, :])**2 / (self.l ** 2)
```

Comment:
Since you are squaring here, I deduce that you do not need the actual distance vectors. That means you should resort to a more efficient way of calculating these. Either you use what exists in sklearn or you write it yourself. The important point is to write out the factors of the binomial. That way you never need the 3rd dimension.